src/cunumeric/sort/sort.cu

```diff
@@ -0,0 +1,742 @@
+  {
+    const size_t num_blocks = (num_splitters + THREADS_PER_BLOCK - 1) / THREADS_PER_BLOCK;
+    extract_split_positions<<<num_blocks, THREADS_PER_BLOCK, 0, stream>>>(
+      local_sorted.values.ptr(0),
+      volume,
+      samples.ptr(0),
+      num_usable_samples,
+      split_positions.ptr(0),
+      num_splitters,
+      my_rank);
+  }
+
+  // need to sync as we share values in between host/device
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+
+  // collect sizes2send, send to rank i: local_sort_data from positions  split_positions[i-1],
+  // split_positions[i] - 1
+  auto size_send = create_buffer<size_t>(num_ranks, Memory::Z_COPY_MEM);
+  {
+    size_t last_position = 0;
+    for (size_t rank = 0; rank < num_ranks - 1; ++rank) {
+      size_t cur_position = split_positions[rank];
+      size_send[rank]     = cur_position - last_position;
+      last_position       = cur_position;
+    }
+    size_send[num_ranks - 1] = volume - last_position;
+  }
+
+  // all2all exchange send/receive sizes
+  auto size_recv = create_buffer<size_t>(num_ranks, Memory::Z_COPY_MEM);
+  CHECK_NCCL(ncclGroupStart());
+  for (size_t r = 0; r < num_ranks; r++) {
+    CHECK_NCCL(ncclSend(size_send.ptr(r), 1, ncclUint64, r, *comm, stream));
+    CHECK_NCCL(ncclRecv(size_recv.ptr(r), 1, ncclUint64, r, *comm, stream));
+  }
+  CHECK_NCCL(ncclGroupEnd());
+
+  // need to sync as we share values in between host/device
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+
+  // allocate merge targets, data transfer...
+  std::vector<SortPiece<VAL>> merge_buffers(num_ranks);
+  std::vector<size_t> aligned_pos_vals_send(num_ranks);
+  std::vector<size_t> aligned_pos_idcs_send(num_ranks);
+
+  size_t buf_size_send_vals_total = 0;
+  size_t buf_size_send_idcs_total = 0;
+  for (size_t i = 0; i < num_ranks; ++i) {
+    // align buffer to allow data transfer of 16byte blocks
+    auto buf_size_vals_recv = get_16b_aligned_count(size_recv[i], sizeof(VAL));
+    merge_buffers[i].values = create_buffer<VAL>(buf_size_vals_recv, Memory::GPU_FB_MEM);
+    merge_buffers[i].size   = size_recv[i];
+
+    aligned_pos_vals_send[i] = buf_size_send_vals_total;
+    buf_size_send_vals_total += get_16b_aligned_count(size_send[i], sizeof(VAL));
+
+    if (argsort) {
+      auto buf_size_idcs_recv  = get_16b_aligned_count(size_recv[i], sizeof(int64_t));
+      merge_buffers[i].indices = create_buffer<int64_t>(buf_size_idcs_recv, Memory::GPU_FB_MEM);
+      aligned_pos_idcs_send[i] = buf_size_send_idcs_total;
+      buf_size_send_idcs_total += get_16b_aligned_count(size_send[i], sizeof(int64_t));
+    } else {
+      merge_buffers[i].indices = create_buffer<int64_t>(0, Memory::GPU_FB_MEM);
+    }
+  }
+
+  // copy values into aligned send buffer
+  auto val_send_buf = local_sorted.values;
+  if (buf_size_send_vals_total > volume) {
+    val_send_buf = create_buffer<VAL>(buf_size_send_vals_total, Memory::GPU_FB_MEM);
```

Comment:
The large aligned buffer is set up in a way that every start position within is 16B aligned. It is basically a concatenation of individual aligned buffers. I put it in one combined chunk in order to simply exchange pointers with the un-aligned data within the transfer method. This way the transfer code is the same in case all data is already aligned in the first place (e.g. for 128bit types).